File: leetcode/1024-video-stitching/solution.py
from typing import List


class Solution:
    def videoStitching(self, clips: List[List[int]], time: int) -> int:
        farthest = [0] * time
        for start, end in clips:
            if start < time:
                farthest[start] = max(farthest[start], end)

        count = 0
        current_end = 0
        next_end = 0
        s = 0
        while current_end < time:
            while s <= current_end and s < time:
                next_end = max(next_end, farthest[s])
                s += 1
            if next_end <= current_end:
                return -1
            count += 1
            current_end = next_end
        return count

File: leetcode/1024-video-stitching/solution.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int videoStitching(vector<vector<int>>& clips, int time) {
        vector<int> farthest(time, 0);
        for (auto& clip : clips) {
            if (clip[0] < time) {
                farthest[clip[0]] = max(farthest[clip[0]], clip[1]);
            }
        }

        int count = 0;
        int currentEnd = 0;
        int nextEnd = 0;
        int s = 0;
        while (currentEnd < time) {
            while (s <= currentEnd && s < time) {
                nextEnd = max(nextEnd, farthest[s]);
                s++;
            }
            if (nextEnd <= currentEnd) {
                return -1;
            }
            count++;
            currentEnd = nextEnd;
        }
        return count;
    }
};
